### ISSUE TITLE
Bitbucket Server as a pipeline configuration source fails

### ISSUE THREAD

**Issue by DuckieHo** [jenkinsci/templating-engine-plugin#61]
**Issue**
When attempting to use "Bitbucket Server" as a source for pipeline configurations, JTE cannot locate the pipeline template.  Testing the connection and other projects using the Bitbucket Server configurtion confirms that Jenkins is able to access the repo.

**Workaround**
Switching to a "Git" as a source, JTE works properly.

**Versions**
Jenkins 2.204.4
Templating Engine 1.5.2
[Bitbucket Server Integration 1.1.0](https://plugins.jenkins.io/atlassian-bitbucket-server-integration/)

**Log**
```
> Running in Durability level: MAX_SURVIVABILITY
[Pipeline] Start of Pipeline
[JTE] Obtained Template Configuration File (show)
[JTE] Obtained Repository Jenkinsfile (show)
[Pipeline] End of Pipeline
org.boozallen.plugins.jte.config.TemplateConfigException: Could not determine pipeline template.
	at sun.reflect.GeneratedConstructorAccessor1376.newInstance(Unknown Source)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.codehaus.groovy.reflection.CachedConstructor.invoke(CachedConstructor.java:83)
	at org.codehaus.groovy.runtime.callsite.ConstructorSite$ConstructorSiteNoUnwrapNoCoerce.callConstructor(ConstructorSite.java:105)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:247)
	at org.boozallen.plugins.jte.TemplateEntryPointVariable.getTemplate(TemplateEntryPointVariable.groovy:193)
	at org.boozallen.plugins.jte.TemplateEntryPointVariable$getTemplate.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at com.cloudbees.groovy.cps.sandbox.DefaultInvoker.methodCall(DefaultInvoker.java:20)
	at script15870935980111752575261.call(script15870935980111752575261.groovy:32)
	at script15870935980111752575261.call(script15870935980111752575261.groovy)
	...
```

**Comment by steven-terrana** [comment]
Thank you for reporting this @DuckieHo. 

## TLDR 

There's a known issue with JTE and BitBucket Cloud caused by [This JIRA Issue](https://jira.atlassian.com/browse/BCLOUD-5814?error=login_required&error_description=Login+required&state=8583653d-0cbf-4ed7-a8a9-987b69956560). 

## Details

The Jenkins Templating Engine relies on the [SCM API 2.0 spec](https://jenkins.io/blog/2017/01/17/scm-api-2/) to fetch files from source code management repositories.  This spec was designed so that consumers of SCM repositories could leverage a common interface.  Unfortunately, because of the JIRA mentioned above, the `bitbucket-branch-source` plugin does not support the SCM API 2.0 spec for an edge case. 

A possible work around for this would be to fall back to doing a generic SCM checkout and then analyzing the repository ourselves instead of relying on the common API. 

## Request for more info 
* Are you using BitBucket Cloud or an on-premises deployment of BitBucket? 
* What job type were you using when this happened? A pipeline job created as part of a Multibranch project? 
* If so - was it a job representing a branch or a pull request? 

Thank you!

**Comment by DuckieHo** [comment]
**_Are you using BitBucket Cloud or an on-premises deployment of BitBucket?_**
BitBucket Server

**_What job type were you using when this happened? A pipeline job created as part of a Multibranch project?_**
If I recall, **pipeline**, **multibranch project**, and **Bitbucket Team/Project** all have this issue.  The issue most likely with the plugin Bitbucket and Bitbucket server plug-ins.  Both plugins are confirming that they can successful connect to the repository.  This causes confusion as while it connect (you think it works) but then JTE throws a misleading error of "cannot find".

_**If so - was it a job representing a branch or a pull request?**_
It was a branch event.

**Comment by steven-terrana** [comment]
Thanks for the additional info. 

I need to do some relationship to figure out the relationship between the [Bitbucket Branch Source](https://plugins.jenkins.io/cloudbees-bitbucket-branch-source/) and the Bitbucket Server Integration plugin

**Comment by DuckieHo** [comment]
History wise?
Atlassian released the "Bitbucket Server Integration" last year. 
The Bitbucket Branch Source is older and supports both cloud/server instance.  It is an orphan now. 
 People made comments on why Atlassian simply did not just adopt it.

**Comment by steven-terrana** [comment]
Interesting. thank you. 

I've opened a [JIRA issue](https://issues.jenkins-ci.org/browse/JENKINS-62002) for the Bitbucket Server Integration plugin to actually implement the SCM API 2.0 spec.. 

In the meantime, as time allows, i will try to make progress on expanding JTE to support any SCM by generically checking out the SCM
